# Proyecto 1
## Sistemas de Datos Intensivos
### Integrantes: Daniel Felipe Ortiz - 202221234, Jenifer Arce 202014993, Gustavo Mendez 202415142

In [0]:
from pyspark.sql import functions as F

# 1. Lea los dos conjuntos de datos. Escriba el código para que la lectura sea lo más rápida posible y justifique su respuesta.

In [0]:
storage_account_name = "uniandesyjt"
container_name = "sid"

secop_path = f"wasbs://{container_name}@{storage_account_name}.blob.core.windows.net/secop"
secop = spark.read.json(secop_path)
bpin_path = f"wasbs://{container_name}@{storage_account_name}.blob.core.windows.net/bpin"
bpin = (
    spark.read
         .option("header", "true")
         .csv(bpin_path)
)

In [0]:
display(secop.head(10))

anno_bpin anno_firma ciudad código_bpin departamento documento_proveedor duración_del_contrato entidad_centralizada estado_contrato fecha_de_fin_del_contrato fecha_de_firma fecha_de_inicio_del_contrato modalidad_de_contratacion objeto_del_contrato orden origen_de_los_recursos proveedor_adjudicado rama sector tipo_de_contrato tipodocproveedor ultima_actualizacion urlproceso valor_del_contrato valor_pagado 2024 2024 No Definido 2023004410229 Huila 1075286176 6 Mes(es) Centralizada En ejecución 2024-09-07 2024-02-28 2024-03-08 Contratación directa PRESTAR SERVICIOS DE APOYO A LA SECRETARIA DE SALUD DEPARTAMENTAL EN LA EJECUCIÓN DE ACTIVIDADES DE ENTORNOS SALUDABLES; INSPECCIÓN SANITARIA Y DE CONTROL DE ENFERMEDADES TRANSMITIDAS POR VECTORES Y ZOONOSIS EN EL ÁREA URBANA Y RURAL DEL MUNICIPIO DE GUADALUPE. Territorial Distribuido Juan David Restrepo Cuellar Ejecutivo Servicio Público Prestación de servicios Cédula de Ciudadanía null https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.5743726&isFromPublicArea=True&isModal=true&asPopupView=true 18900000 $2,520,000 2024 2024 Cartagena 2024130010043 Bolívar 901453530 18 Dia(s) Centralizada En ejecución 2024-12-31 2024-12-12 2024-12-16 Mínima cuantía CONTRATAR POR MONTO AGOTABLE LA ADQUISICIÓN DE KITS DE ASEO
PERSONAL CON DESTINO A LAS PERSONAS PRIVADAS DE LA LIBERTAD RECLUIDAS EN LA
CIUDAD DE CARTAGENA DE INDIAS Territorial Distribuido TECNOSUMINISTROS FMK S.A.S Ejecutivo No aplica/No pertenece Suministros NIT null https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.7148502&isFromPublicArea=True&isModal=true&asPopupView=true 120000000 $0 2024 2024 Girardot 2024253070020 Cundinamarca 1070604592 3 Mes(es) Centralizada Modificado 2024-12-05 2024-09-04 2024-09-06 Contratación directa PRESTACION DE SERVICIOS PROFESIONALES DE UN (A) TRABAJADOR (A) SOCIAL PARA EL ACOMPAÑAMIENTO PSICOSOCIAL AL PROGRAMA COLOMBIA MAYOR LIDERADO POR LA SECRETARIA DE DESARROLLO ECONOMICO Y SOCIAL Territorial Distribuido JINETH TATIANA GOMEZ IBAÑEZ Ejecutivo Servicio Público Prestación de servicios Cédula de Ciudadanía 2024 Nov 01 12:00:00 AM https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6652481&isFromPublicArea=True&isModal=true&asPopupView=true 9000000 $9,000,000 2024 2024 Puerto Gaitán 2024505680033 Meta 94528216 10 Dia(s) Descentralizada terminado 2024-12-27 2024-12-16 2024-12-17 Mínima cuantía ADQUISICIÓN E INSTALACIÓN DE CAMPAÑAS VISUALES DE PUBLICIDAD CON RELACIÓN A LA PREVENCIÓN DE LOS DIFERENTES ESCENARIOS DE RIESGO Y/O EMERGENCIA QUE SE PUEDEN PRESENTAR EN EL MUNICIPIO DE PUERTO GAITÁN META Territorial Distribuido JOSE WALTER ASTUDILLO SUAREZ Ejecutivo Servicio Público Compraventa Cédula de Ciudadanía 2025 Mar 05 12:00:00 AM https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.7185864&isFromPublicArea=True&isModal=true&asPopupView=true 67599015 $67,599,015 No D 2024 Medellín No Definido Antioquia 1007461247 165 Dia(s) Centralizada En ejecución 2024-12-30 2024-07-12 2024-07-16 Contratación directa Prestación de servicios como contratista independiente; sin vínculo laboral por su propia cuenta y riesgo para realizar la gestion de Apoyo administrativo para la gestión documental en ejecución del Contrato Interadministrativo No.4600100862 DE 2024; celebrado entre EL DISTRITO ESPECIAL DE CIENCIA TECNOLOGÍA E INNOVACIÓN DE MEDELLÍN;SECRETARÍA DE INNOVACION y el ITM Nacional Recursos Propios Maria camila Jimenez Gomez Corporación Autónoma Educación Nacional Prestación de servicios Cédula de Ciudadanía null https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6396174&isFromPublicArea=True&isModal=true&asPopupView=true 11618448 $0 2023 2024 Funza 2022252860043 Cundinamarca 1073507910 10 Mes(es) Centralizada Cerrado 2024-12-26 2024-02-26 2024-02-27 Contratación directa PRESTACION DE SERVICIOS PROFESIONALES COMO ABOGADO PARA B

In [0]:
display(bpin.head(10))

Bpin NombreProyecto ObjetivoGeneral EstadoProyecto Horizonte Sector EntidadResponsable ProgramaPresupuestal TipoProyecto PlanDesarrolloNacional ValorTotalProyecto ValorVigenteProyecto ValorObligacionProyecto ValorPagoProyecto SubEstadoProyecto CodigoEntidadResponsable TotalBeneficiario 2019415240011 Desarrollo Del Programa De Transporte Escolar En El Municipio De Palermo Huila Garantizar a los menores y los adolescentes del municipio de Palermo su derecho Constitucional de acceder a educación pública. En Ejecución 2019-2019 Educación PalermoHuila 2201 CALIDAD, COBERTURA Y FORTALECIMIENTO DE LA EDUCACIÓN INICIAL, PRESCOLAR, BÁSICA Y MEDIA T Todos por un nuevo país (2014-2018) 1000000000.0 1000000000.0 0.0 0.0 Inactivo  (PGN, Territorio) 41524 718 2022002700139 Desarrollo De La Dimensión Vida Saludable Y Enfermedades Transmisibles - Enfermedades Emergentes Reemergentes Y Desatendidas Del Programa Salud Pública Prioritaria Y Diferente Vigencia 2023 En El Departamento De Sucre Mejorar las condiciones de atención y protección en salud a la población del departamento de Sucre. En Ejecución 2023-2023 Salud Y Protección Social Sucre SALUD PÚBLICA T Pacto Por Colombia, Pacto por la Equidad (2019-2022) 370066980.0 370066980.0 106499940.0 106499940.0 Inactivo  (PGN, Territorio) 70 981727 2024735550002 Mejoramiento De Las Condiones De Acciones Y Acceso Y Permanencia En El Sistema Educativo De La Población Estudiantil Del Resguardo Indigena Paez Nasa Wesx Del Corregimiento De Gaitania Municipio De Planadas Tolima GTYBY Formulación 2024-2024 Educación PlanadasTolima 2201 CALIDAD, COBERTURA Y FORTALECIMIENTO DE LA EDUCACIÓN INICIAL, PRESCOLAR, BÁSICA Y MEDIA T Colombia Potencia Mundial de la Vida (2022-2026) 1000.0 0.0 0.0 0.0 Devuelto a MGA 73555 100 2024202380105 Desarrollo Y Fomento De La Cultura Local Del Municipio El Copey Fortalecer la promoción de actividades artísticas y culturales en el municipio de El Copey Cesar En Ejecución 2024-2024 Cultura El CopeyCesar 3301 PROMOCIÓN Y ACCESO EFECTIVO A PROCESOS CULTURALES Y ARTÍSTICOS T Colombia Potencia Mundial de la Vida (2022-2026) 469820299.0 469820299.0 0.0 0.0 Inactivo  (PGN, Territorio) 20238 32624 2019522870015 Suministro De Materiales De Ferretería Y De Cantera Para La Adecuación De Varios Puntos Álgidos No Continuos Sobre Las Vías Que Conducen A La Soledad Y Guapuscal Bajo Que Necesitan Ser Intervenidos Para El Transito Seguro En El Municipio De Funes Mejorar la infraestructura de las vías del sector rural Formulación 2019-2019 Transporte FunesNariño 2401 INFRAESTRUCTURA RED VIAL PRIMARIA T Todos por un nuevo país (2014-2018) 68796600.0 0.0 0.0 0.0 Devuelto a MGA 52287 4500 2022862190016 Adecuación De Las Instalaciones Del Estadio Jesus Maria Cabal En El Muicipio De Colon Colon ESTIMULAR Y PROMOVER LA ACTIVIDAD FÍSICA Y EL DEPORTE PARA LA ORGANIZACIÓN Y PREMIACIÓN DEL FESTIVAL DE ESCUELAS DEL VALLE DE SIBUNDOY EN CONMEMORACIÓN A LAS FIESTAS PATRONALES SAN ANTONIO DE PADUA COLON En Ejecución 2022-2022 Deporte Y Recreación ColónPutumayo 4301 FOMENTO A LA RECREACIÓN, LA ACTIVIDAD FÍSICA Y EL DEPORTE PARA DESARROLLAR ENTORNOS DE CONVIVENCIA Y PAZ T Pacto Por Colombia, Pacto por la Equidad (2019-2022) 15346286.8 15346286.8 0.0 0.0 Inactivo  (PGN, Territorio) 86219 5530 2021760540027 Desarrollo Del Programa Unidos Haremos Más Desarrollo Social - Sector Cultura Del Municipio De Argelia Valle Del Cauca insuficientes programas y actividades dirigidos al sector cultura del municipio de Argelia valle En Ejecución 2022-2022 Cultura ArgeliaValle del Cauca 3301 PROMOCIÓN Y ACCESO EFECTIVO A PROCESOS CULTURALES Y ARTÍSTICOS T Pacto Por Colombia, Pacto por la Equidad (2019-2022) 160291016.22 160291016.22 72050446.0 52310000.0 Inactivo  (PGN, Territorio) 76054 1809 2018235700040 Reposicin De Redes De Alcantarillado En La Transversal 11 Entre Calles 17a Y Calle 18b Entre Barrio El Carmen Y Av. Jose Ines Torres Del Municipio De Pueblo Nuevo Departamento De Cordoba Disminuir los niveles de disposicin d

In [0]:
secop.printSchema()

root
 |-- anno_bpin: string (nullable = true)
 |-- anno_firma: integer (nullable = true)
 |-- ciudad: string (nullable = true)
 |-- código_bpin: string (nullable = true)
 |-- departamento: string (nullable = true)
 |-- documento_proveedor: string (nullable = true)
 |-- duración_del_contrato: string (nullable = true)
 |-- entidad_centralizada: string (nullable = true)
 |-- estado_contrato: string (nullable = true)
 |-- fecha_de_fin_del_contrato: string (nullable = true)
 |-- fecha_de_firma: string (nullable = true)
 |-- fecha_de_inicio_del_contrato: string (nullable = true)
 |-- modalidad_de_contratacion: string (nullable = true)
 |-- objeto_del_contrato: string (nullable = true)
 |-- orden: string (nullable = true)
 |-- origen_de_los_recursos: string (nullable = true)
 |-- proveedor_adjudicado: string (nullable = true)
 |-- rama: string (nullable = true)
 |-- sector: string (nullable = true)
 |-- tipo_de_contrato: string (nullable = true)
 |-- tipodocproveedor: string (nullable = true)

In [0]:
bpin.printSchema()

root
 |-- Bpin: string (nullable = true)
 |-- NombreProyecto: string (nullable = true)
 |-- ObjetivoGeneral: string (nullable = true)
 |-- EstadoProyecto: string (nullable = true)
 |-- Horizonte: string (nullable = true)
 |-- Sector: string (nullable = true)
 |-- EntidadResponsable: string (nullable = true)
 |-- ProgramaPresupuestal: string (nullable = true)
 |-- TipoProyecto: string (nullable = true)
 |-- PlanDesarrolloNacional: string (nullable = true)
 |-- ValorTotalProyecto: string (nullable = true)
 |-- ValorVigenteProyecto: string (nullable = true)
 |-- ValorObligacionProyecto: string (nullable = true)
 |-- ValorPagoProyecto: string (nullable = true)
 |-- SubEstadoProyecto: string (nullable = true)
 |-- CodigoEntidadResponsable: string (nullable = true)
 |-- TotalBeneficiario: string (nullable = true)



# 2. Identifique los 10 **proveedores** que han tenido el mayor **valor de contratos** durante el año **2024**

Esta sección calcula el top 10 de proveedores en 2024 según el mayor valor de un contrato individual firmado por cada proveedor (no la suma de todos sus contratos).

## Qué hace el código (siguiente celda)
1) Filtra los registros de SECOP cuyo año de `fecha_de_inicio_del_contrato` sea 2024.
2) Agrupa por proveedor usando `documento_proveedor`.
3) Calcula, para cada proveedor, el valor máximo de contrato con `max(valor_del_contrato)` y lo etiqueta como `maximo_valor_de_contrato`.
4) Ordena de forma descendente por ese valor máximo y toma los 10 primeros.
5) Visualiza el resultado.

## Datos relevantes
- **proveedor**: `secop.documento_proveedor`
- **valor de contratos** (considerado como máximo por proveedor): `secop.valor_del_contrato`
- **año** (del contrato): `secop.fecha_de_inicio_del_contrato` (se extrae el año con `year(...)`)


In [0]:
top10 = (
    secop.filter(F.year("fecha_de_inicio_del_contrato") == 2024)
        .groupBy("documento_proveedor")
        .agg(F.max("valor_del_contrato").alias("maximo_valor_de_contrato"))
        .orderBy(F.desc("maximo_valor_de_contrato"))
        .limit(10)
)

display(top10)

documento_proveedor maximo_valor_de_contrato 30310147 2115000000000 899999162 574625385908 No Definido 556028077752 890922447 540544008112 830144890 533333333333 800096329 520008781105 900265408 500000000000 800213678 481658137903 9017853484 469047000000 9017774360 446445908445

# 3. Identifique los 10 Proyectos de inversión con mayor valor de contratos sin pagar

A continuación se muestran las celdas que calculan, para cada contrato, el valor sin pagar y listan las 10 filas con mayor valor pendiente, junto con su proveedor y el nombre del proyecto asociado.

## Qué hacen las siguientes celdas

1) Limpieza y tipado de pagos:
   - Se crea la columna `valor_pagado_num` a partir de `valor_pagado`, eliminando símbolos de moneda y comas (`$`, `,`) y convirtiéndola a tipo numérico (double).
   - Si `valor_pagado` está vacío o mal formado, el resultado será `null`.

2) Unión con BPIN:
   - Se realiza un `inner join` entre `secop` y `bpin` usando la llave foránea `secop.código_bpin == bpin.Bpin` para anexar el `NombreProyecto` al contrato.

3) Cálculo del valor sin pagar:
   - Se crea la columna `valor_sin_pagar = valor_del_contrato - valor_pagado_num`.

4) Selección, orden y top 10:
   - Se seleccionan las columnas `documento_proveedor`, `valor_sin_pagar` y `NombreProyecto`.
   - Se ordena de manera descendente por `valor_sin_pagar`.
   - Se limita a las 10 filas con mayor valor pendiente y se visualiza el resultado.

Nota: El listado es a nivel de contrato (no agrega por proyecto); por ello se muestran los contratos con mayor saldo pendiente, incluyendo su proveedor y el nombre del proyecto.


In [0]:
secop = secop.withColumn(
    "valor_pagado_num",
    F.regexp_replace("valor_pagado", "[$,]", "").cast("double")
)

In [0]:
joined_inner = (
    secop.join(bpin, secop.código_bpin == bpin.Bpin, "inner")
)

In [0]:
top10 = (
    joined_inner
    .withColumn("valor_sin_pagar", F.col("valor_del_contrato") - F.col("valor_pagado_num"))
    .select("documento_proveedor", "valor_sin_pagar", "NombreProyecto")
    .orderBy(F.desc("valor_sin_pagar"))
    .limit(10)
)

display(top10)

documento_proveedor valor_sin_pagar NombreProyecto 74372236 2.3856178E12 Apoyo Para La Gestión Eficiente Del Recurso Hídrico En El Sector Agropecuario Mediante La Séptima Convocatoria Del Fondo Finca En El Departamento De Boyacá 900623766 1.140538892678E12 Construcción Y Operación De La Cárcel Metropolitana Para Sindicados Medellín 900118630 5.39075229979E11 Fortalecimiento De Los Procesos De Aseguramiento Y Prestación Integral De Servicios De Salud Con Calidad Inclusión Y Equidad Para Mejorar La Gobernanza Del Sistema General De Seguridad Social En Salud En El Departamento Risaralda 800215293 5.11875E11 Fortalecimiento De La Participación E Inclusión De La Población En Recreación Actividad Física Deporte Formativo Y Social Comunitario En El Valle Del Cauca. Valle Del Cauca 900265408 5.0E11 Implementacion Del Ordenamiento Social De La Propiedad Rural A Nivel Nacional 900265408 4.74031380398E11 Implementacion Del Ordenamiento Social De La Propiedad Rural A Nivel Nacional 800096329 4.7E11 Mejoramiento De Las Condiciones De Infraestructura De Las Instituciones De Educación Superior Públicas Nacional 900249143 4.4390691E11 Diseño Gestión Promoción E Implementación Del Plan Maestro Para La Conservación Integral Revitalización Y Sostenibilidad De Los Proyectos De Infraestructura Turística Del Departamento Del Atlántico 901907544 3.84542139555E11 Implementación De La Estrategia Alimentando Sueños Y Conocimientos Alimentación Escolar Cartagena De Indias 800116398 3.79892382549E11 Servicio Financiero Y Gestión Del Riesgo Para El Sector Agropecuario Y Rural Sostenible La Agroindustrialización Y La Producción Agroalimentaria Nacional

# 4) Identifique para cada año el top 5 de los proveedores con mayor valor de contratos. La respuesta debe incluir el nombre de los proveedores, el valor de sus contratos y su posición en el top. 

En este ejercicio lo que se hizo fue primero hacer un groupby por el año de firma junto por el proveedor, de esta manera se conseguía el valor de todos los contratos de un proveedor en un año en concreto. A partir de esto se suman los valores de los contratos. El resultado de este primer paso es un dataframe que se llama **proveedor_anno_df**.

In [0]:
from pyspark.sql import Window

proveedor_anno_df = (
    secop.groupBy("anno_firma", "proveedor_adjudicado")
         .agg(F.sum("valor_del_contrato").alias("total_valor_contrato"))
)

display(proveedor_anno_df.head(5))

anno_firma proveedor_adjudicado total_valor_contrato 2024 Zaray Valentina Diaz Alba 8100000 2024 PEDRO VENANCIO MANJARREZ CABANA 115500000 2024 LUIS JAVIER MORALES GARCIA 38871261 2024 DISUR SAS 7513414282 2024 ELKIN ARLEY FRANCO MARULANDA 21483000

A partir de lo anterior se realiza un partionby con una window, esto para poder segmentar los datos por el año de firma, de esta manera se puede calcular los rankings por cada año (columna ranking_anno). Después de esto, se realiza una operación de **rank** para poder rankear a los proveedores por año de acuerdo al total de los valores que tengan, después de esto se realiza un ordenamiento descendiente. Este dense_rank se hace sobre la columna que se hizo anteriormente en el partition, el nombre es: ranking_anno. Finalmente el dataframe **top5_df** será la respuesta a la pregunta del enunciado. En total hay 55 filas, porque son varios años a analizar, por ejemplo para el año 2016 los proveedores con más dinero de contratos son:
1. KAPITAL
2. John Alexander Ardila Evan
3. LUZ ESTELA ROJAS QUINTERO
4. UNION TEMPORAL GENERALI COLOMBIA SEGUROS GENERALES S.A - AIG SEGUROS COLOMBIA S.A.- ACE SEGUROS S.A- CHUBB DE COLOMBIA COMPAÑÍA DE SEGUROS S.A- CHUBB
5. AGUILA DE ORO DE COLOMBIA LTDA

fuente:
https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.functions.rank.html

In [0]:
ranking_anno = Window.partitionBy("anno_firma").orderBy(F.desc("total_valor_contrato"))
top5_df = (
    proveedor_anno_df.withColumn("posicion", F.rank().over(ranking_anno))
          .filter(F.col("posicion") <= 5)
)

display(top5_df)

anno_firma proveedor_adjudicado total_valor_contrato posicion 2015 UNIÓN TEMPORAL IDOM-SILO 563760000 1 2015 CORPORACIÓN ECOVERSA 520000000 2 2015 UT MAPFRE - PREVISORA 145120101 3 2015 GRUPO SPIRA SAS 119828000 4 2015 AXA COLPATRIA SEGUROS S.A 98600000 5 2016 KAPITAL 151889229120 1 2016 John Alexander Ardila Evan 27000000000 2 2016 LUZ ESTELA ROJAS QUINTERO 18000000000 3 2016 UNION TEMPORAL GENERALI COLOMBIA SEGUROS GENERALES S.A - AIG SEGUROS COLOMBIA S.A.- ACE SEGUROS S.A- CHUBB DE COLOMBIA COMPAÑÍA DE SEGUROS S.A- CHUBB 17958407612 4 2016 AGUILA DE ORO DE COLOMBIA LTDA 14031229446 5 2017 FESU 438187721616 1 2017 COLFONDOS SA 311718748000 2 2017 ESTACION DE SERVICIO CARRERA 50 SAS 211108041588 3 2017 CAJA DE COMPENSACION FAMILIAR COMPENSAR 193694066424 4 2017 ENTerritorio S.A 159258068054 5 2018 ALAN ARCHBOLD MARTINEZ 4330119839236 1 2018 SOCIEDAD TEQUENDAMA S.A. 1907805561976 2 2018 LIMPIEZA METROPOLITANA S A ESP 1544233498292 3 2018 Promoambiental Distrito S.A.S. ESP 1101109931961 4 2018 UNION TEMPORAL MEDIPOL 16 988378423013 5 2019 FRANCISCO ANTONIO ALVARADO BESTENE 9645115773936 1 2019 MARTHA LEON SANCHEZ 2715992374000 2 2019 INSTITUTO COLOMBIANO DE CREDITO EDUCATIVO Y ESTUDIO TÉCNICOS EN EL EXTERIOR MARIANO OSPINA PEREZ. 1737035145293 3 2019 Colombia Telecomunicaciones S.A. ESP BIC 1114488772821 4 2019 MAPFRE SEGUROS GENERALES DE COLOMBIA S.A. 484970594047 5 2020 JOHN ESTIVEN MARIN RODRIGUEZ 4949784128152 1 2020 ALBA MILENA FAJARDO ISAZA 4949777903187 2 2020 ESE HOSPITAL UNIVERSITARIO JULIO MENDEZ BARRENECHE 1574884067000 3 2020 FINANCIERA DE DESARROLLO TERRITORIAL S.A. 1163497454841 4 2020 CONSORCIO EUCARÍSTICO CARRERA 68 1162759703824 5 2021 CUERPO DE BOMBEROS VOLUNTARIOS DE ITAGUI 4092141470866 1 2021 AHMSA 3482943972000 2 2021 ENTerritorio S.A 3157488993927 3 2021 FINANCIERA DE DESARROLLO TERRITORIAL S.A. 1947916672054 4 2021 UNION TEMPORAL DISTRIBUCION PROCESOS ELECTORALES 2021 1799707249746 5 2022 Corporacion de Ciencia y Tecnologia para el Desarrollo de la Industria Naval, Maritima y Fluvial 2363616887571 1 2022 INGREDION COLOMBIA SA 1154999740390 2 2022 UNIVERSIDAD DE ANTIOQUIA 1130253375440 3 2022 UT ETB NET COLOMBIA CONECTADA 1072552301475 4 2022 FINANCIERA DE DESARROLLO TERRITORIAL S.A. 905428602305 5 2023 FINANCIERA DE DESARROLLO TERRITORIAL S.A. 3104985006258 1 2023 UNIVERSIDAD DE ANTIOQUIA 1615818359492 2 2023 FONDO MIXTO PARA LA PROMOCION DE LA CULTURA Y LAS ARTES DE BOYACA 1236000209976 3 2023 AGENCIA LOGISTICA DE LAS FUERZAS MILITARES 1170044220037 4 2023 PROGRAMA DE LAS NACIONES UNIDAS PARA EL DESARROLLO PNUD 1150227507353 5 2024 MARIA SORELY GRISALES / CREACIONES BJ 2115156861243 1 2024 UNIVERSIDAD DE ANTIOQUIA 1856417293206 2 2024 INSTITUTO PARA EL DESARROLLO DE ANTIOQUIA IDEA 1703797668980 3 2024 FINANCIERA DE DESARROLLO TERRITORIAL S.A. 1541604758133 4 2024 INSTITUTO TECNOLOGICO METROPOLITANO 1152177220033 5 2025 Zona Franca Barranquilla 2846224257835 1 2025 Yesid Avila Torres 2385617800000 2 2025 ESU 2200716981700 3 2025 UNION TEMPORAL INTEGRACION LOGISTICA ELECTORAL 2026 2124806917129 4 2025 INSTITUTO PARA EL DESARROLLO DE ANTIOQUIA IDEA 1413834935935 5

# 5) Identifique el número de **proveedores** que firmaron contratos en el **2024** y **NO** firmaron contratos en el **2020**.

## Datos relevantes
- **Proveedor**: secop.documento_proveedor
- **Año de Firma**: secop.anno_firma


En este problema se crean dos DataFrames filtrando los contratos por los años 2024 y 2020.
Después de aplicar los filtros y seleccionar únicamente la columna documento_proveedor, se eliminan duplicados para conservar proveedores únicos.
Luego se realiza un left_anti join, restando a los proveedores de 2024 aquellos que también aparecen en 2020.
Finalmente, se cuenta el número de filas resultantes, que corresponde a la cantidad de proveedores que firmaron contratos en 2024 y no firmaron en 2020.

In [0]:
filtro24 = (secop
            .filter(secop["anno_firma"]==2024)
            .select("documento_proveedor")
            .distinct()
)

filtro20 = (secop
            .filter(secop["anno_firma"]==2020)
            .select("documento_proveedor")
            .distinct()
)

numeroProvedores = filtro24.join(filtro20, "documento_proveedor", "left_anti").count()
print(numeroProvedores)


394208


# 6) Calcule el valor promedio de los contratos de “Prestación de servicios” y una aproximación de la mediana (el percentil 50). Construya esta respuesta de manera que se minimice el tiempo de cálculo, aunque se sacrifique precisión. (13%) 

El ejercicio lo vamos a comenzar filtrando solo los contratos por prestación de servicios:

In [0]:
prestacion_df = secop.filter(F.col("tipo_de_contrato") == "Prestación de servicios")
display(prestacion_df.head(3))
print("df original filas:", secop.count(), "prestacion_df:", prestacion_df.count())

anno_bpin anno_firma ciudad código_bpin departamento documento_proveedor duración_del_contrato entidad_centralizada estado_contrato fecha_de_fin_del_contrato fecha_de_firma fecha_de_inicio_del_contrato modalidad_de_contratacion objeto_del_contrato orden origen_de_los_recursos proveedor_adjudicado rama sector tipo_de_contrato tipodocproveedor ultima_actualizacion urlproceso valor_del_contrato valor_pagado valor_pagado_num 2024 2024 No Definido 2023004410229 Huila 1075286176 6 Mes(es) Centralizada En ejecución 2024-09-07 2024-02-28 2024-03-08 Contratación directa PRESTAR SERVICIOS DE APOYO A LA SECRETARIA DE SALUD DEPARTAMENTAL EN LA EJECUCIÓN DE ACTIVIDADES DE ENTORNOS SALUDABLES; INSPECCIÓN SANITARIA Y DE CONTROL DE ENFERMEDADES TRANSMITIDAS POR VECTORES Y ZOONOSIS EN EL ÁREA URBANA Y RURAL DEL MUNICIPIO DE GUADALUPE. Territorial Distribuido Juan David Restrepo Cuellar Ejecutivo Servicio Público Prestación de servicios Cédula de Ciudadanía null https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.5743726&isFromPublicArea=True&isModal=true&asPopupView=true 18900000 $2,520,000 2520000.0 2024 2024 Girardot 2024253070020 Cundinamarca 1070604592 3 Mes(es) Centralizada Modificado 2024-12-05 2024-09-04 2024-09-06 Contratación directa PRESTACION DE SERVICIOS PROFESIONALES DE UN (A) TRABAJADOR (A) SOCIAL PARA EL ACOMPAÑAMIENTO PSICOSOCIAL AL PROGRAMA COLOMBIA MAYOR LIDERADO POR LA SECRETARIA DE DESARROLLO ECONOMICO Y SOCIAL Territorial Distribuido JINETH TATIANA GOMEZ IBAÑEZ Ejecutivo Servicio Público Prestación de servicios Cédula de Ciudadanía 2024 Nov 01 12:00:00 AM https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6652481&isFromPublicArea=True&isModal=true&asPopupView=true 9000000 $9,000,000 9000000.0 No D 2024 Medellín No Definido Antioquia 1007461247 165 Dia(s) Centralizada En ejecución 2024-12-30 2024-07-12 2024-07-16 Contratación directa Prestación de servicios como contratista independiente; sin vínculo laboral por su propia cuenta y riesgo para realizar la gestion de Apoyo administrativo para la gestión documental en ejecución del Contrato Interadministrativo No.4600100862 DE 2024; celebrado entre EL DISTRITO ESPECIAL DE CIENCIA TECNOLOGÍA E INNOVACIÓN DE MEDELLÍN;SECRETARÍA DE INNOVACION y el ITM Nacional Recursos Propios Maria camila Jimenez Gomez Corporación Autónoma Educación Nacional Prestación de servicios Cédula de Ciudadanía null https://community.secop.gov.co/Public/Tendering/OpportunityDetail/Index?noticeUID=CO1.NTC.6396174&isFromPublicArea=True&isModal=true&asPopupView=true 11618448 $0 0.0

df original filas: 4447723 prestacion_df: 3813168


Para encontrar la mediana se puede hacer de la manera clásica que sería ordenar todo el dataset, y después tomar la mediana que es la mitad de los datos. Sin embargo, de esta manera el tiempo de cálculo incrementa, ya que se tendría que ordenar 3.813.168 datos. Al revisar la documentación de pyspark hay un método que se llama approxQuantile, que permite calcular cuartiles de una manera aproximada, es decir que usando este método se conseguiría minimizar el tiempo de cálculo (pero se sacrifica precisión). Por lo tanto, a continuación se calculará la mediana:

Fuente:
https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/api/pyspark.sql.DataFrame.approxQuantile.html

In [0]:
mediana_approxquantile = prestacion_df.approxQuantile("valor_del_contrato", [0.5], 0.05)[0] 
print("Mediana aproximada con approxQuantile de contratos con prestación de servicios:" , mediana_approxquantile) 

Mediana aproximada con approxQuantile de contratos con prestación de servicios: 17764705.0


Al comparar el tiempo con un método tradicional para encontrar la mediana (hacer primero order by, y después conseguir la mitad de los datos):

In [0]:
from pyspark.sql import Window

num_filas = prestacion_df.count()

# pos central (par/impar)
k1 = num_filas//2 + 1
k2 = (num_filas+1)//2

prestacion_df_pos = prestacion_df.withColumn("pos", F.row_number().over(Window.orderBy("valor_del_contrato")))
med_exact = prestacion_df_pos.filter((F.col("pos") == k1)  | (F.col("pos") == k2))
display(med_exact.select("valor_del_contrato", "pos"))

/databricks/python/lib/python3.12/site-packages/pyspark/sql/connect/expressions.py:1061: UserWarning: WARN WindowExpression: No Partition Defined for Window operation! Moving all data to a single partition, this can cause serious performance degradation.
  warnings.warn(


valor_del_contrato pos 18702456 1906584 18702456 1906585

Para el promedio se utilizará la función agregada avg

In [0]:
promedio = prestacion_df.agg(F.avg("valor_del_contrato").alias("promedio"))
display(promedio.select("promedio"))

promedio 6.948853764571217E7

## Conclusiones punto 6:
- El valor de la mediana con approxQuantile es 17764705, esta es la manera de hacerlo más rápido (sacrificamos precisión).
- El valor real de la mediana (manera tradicional) es 18702456.
- Al ejecutar con approxQuantile el tiempo de ejecución fue 3.8s, mientras que de la manera exacta son 7s.
- El promedio se calculó con avg, el valor es 69.488.537.

# 7) Cuál fue el **año** en el que firmaron contratos un **mayor** número de proveedores distintos.
### Construya esta respuesta de manera que se minimice el tiempo de cálculo, aunque se sacrifique precisión.

## Datos relevantes
- **Proveedor**: secop.documento_proveedor
- **Año de Firma**: secop.anno_firma


Se seleccionan las columnas anno_firma y documento_proveedor, eliminando registros nulos.
Luego se agrupa por año (anno_firma) y se aplica la función approx_count_distinct sobre la columna documento_proveedor para contar proveedores únicos de manera aproximada y eficiente.
Después, se ordenan los resultados de forma descendente por el total de proveedores y se usa limit(1) para quedarse con el año que tiene el mayor número de proveedores distintos.

In [0]:
mayorAnio = (secop
       .select("anno_firma", "documento_proveedor")
       .dropna()
       .groupBy("anno_firma")
       .agg(F.approx_count_distinct("documento_proveedor").alias("Total"))
       .orderBy(F.desc("Total"))
       .limit(1)
)
display(mayorAnio)

anno_firma Total 2024 556412

# 8) Cuáles son las **20 palabras** más comunes en los objetos contractuales del **20%** de los contratos más altos del **2020**. Para esta respuesta omita stopwords puntuaciones y no diferencie entre mayúsculas y minúsculas.

## Datos relevantes
- **Valor de contratos**: secop.valor_del_contrato
- **Año de Firma**: secop.anno_firma
- **Objeto del Contrato**: secop.objeto_del_contrato

Se filtran los contratos del año 2020 y se ordenan de manera descendente por el valor del contrato.
A partir de este conjunto, se selecciona el 20% superior de contratos según su valor.
Sobre la columna objeto_del_contrato, se normaliza el texto a minúsculas, se eliminan signos de puntuación y se tokeniza en palabras.
Se eliminan las stopwords y las cadenas vacías, y luego se cuentan las ocurrencias de cada palabra.
Finalmente, se ordenan los resultados en forma descendente y se seleccionan las 20 palabras más frecuentes.

In [0]:
stopwords = ["a", "actualmente", "adelante", "además", "afirmó", "agregó", "ahora", "ahí", "al", "algo", "alguna", "algunas", "alguno", "algunos", "algún", "alrededor", "ambos", "ampleamos", "ante", "anterior", "antes", "apenas", "aproximadamente", "aquel", "aquellas", "aquellos", "aqui", "aquí", "arriba", "aseguró", "así", "atras", "aunque", "ayer", "añadió", "aún", "bajo", "bastante", "bien", "buen", "buena", "buenas", "bueno", "buenos", "cada", "casi", "cerca", "cierta", "ciertas", "cierto", "ciertos", "cinco", "comentó", "como", "con", "conocer", "conseguimos", "conseguir", "considera", "consideró", "consigo", "consigue", "consiguen", "consigues", "contra", "cosas", "creo", "cual", "cuales", "cualquier", "cuando", "cuanto", "cuatro", "cuenta", "cómo", "da", "dado", "dan", "dar", "de", "debe", "deben", "debido", "decir", "dejó", "del", "demás", "dentro", "desde", "después", "dice", "dicen", "dicho", "dieron", "diferente", "diferentes", "dijeron", "dijo", "dio", "donde", "dos", "durante", "e", "ejemplo", "el", "ella", "ellas", "ello", "ellos", "embargo", "empleais", "emplean", "emplear", "empleas", "empleo", "en", "encima", "encuentra", "entonces", "entre", "era", "erais", "eramos", "eran", "eras", "eres", "es", "esa", "esas", "ese", "eso", "esos", "esta", "estaba", "estabais", "estaban", "estabas", "estad", "estada", "estadas", "estado", "estados", "estais", "estamos", "estan", "estando", "estar", "estaremos", "estará", "estarán", "estarás", "estaré", "estaréis", "estaría", "estaríais", "estaríamos", "estarían", "estarías", "estas", "este", "estemos", "esto", "estos", "estoy", "estuve", "estuviera", "estuvierais", "estuvieran", "estuvieras", "estuvieron", "estuviese", "estuvieseis", "estuviesen", "estuvieses", "estuvimos", "estuviste", "estuvisteis", "estuviéramos", "estuviésemos", "estuvo", "está", "estábamos", "estáis", "están", "estás", "esté", "estéis", "estén", "estés", "ex", "existe", "existen", "explicó", "expresó", "fin", "fue", "fuera", "fuerais", "fueran", "fueras", "fueron", "fuese", "fueseis", "fuesen", "fueses", "fui", "fuimos", "fuiste", "fuisteis", "fuéramos", "fuésemos", "gran", "grandes", "gueno", "ha", "haber", "habida", "habidas", "habido", "habidos", "habiendo", "habremos", "habrá", "habrán", "habrás", "habré", "habréis", "habría", "habríais", "habríamos", "habrían", "habrías", "habéis", "había", "habíais", "habíamos", "habían", "habías", "hace", "haceis", "hacemos", "hacen", "hacer", "hacerlo", "haces", "hacia", "haciendo", "hago", "han", "has", "hasta", "hay", "haya", "hayamos", "hayan", "hayas", "hayáis", "he", "hecho", "hemos", "hicieron", "hizo", "hoy", "hube", "hubiera", "hubierais", "hubieran", "hubieras", "hubieron", "hubiese", "hubieseis", "hubiesen", "hubieses", "hubimos", "hubiste", "hubisteis", "hubiéramos", "hubiésemos", "hubo", "igual", "incluso", "indicó", "informó", "intenta", "intentais", "intentamos", "intentan", "intentar", "intentas", "intento", "ir", "junto", "la", "lado", "largo", "las", "le", "les", "llegó", "lleva", "llevar", "lo", "los", "luego", "lugar", "manera", "manifestó", "mayor", "me", "mediante", "mejor", "mencionó", "menos", "mi", "mientras", "mio", "mis", "misma", "mismas", "mismo", "mismos", "modo", "momento", "mucha", "muchas", "mucho", "muchos", "muy", "más", "mí", "mía", "mías", "mío", "míos", "nada", "nadie", "ni", "ninguna", "ningunas", "ninguno", "ningunos", "ningún", "no", "nos", "nosotras", "nosotros", "nuestra", "nuestras", "nuestro", "nuestros", "nueva", "nuevas", "nuevo", "nuevos", "nunca", "o", "ocho", "os", "otra", "otras", "otro", "otros", "para", "parece", "parte", "partir", "pasada", "pasado", "pero", "pesar", "poca", "pocas", "poco", "pocos", "podeis", "podemos", "poder", "podria", "podriais", "podriamos", "podrian", "podrias", "podrá", "podrán", "podría", "podrían", "poner", "por", "por qué", "porque", "posible", "primer", "primera", "primero", "primeros", "principalmente", "propia", "propias", "propio", "propios", "próximo", "próximos", "pudo", "pueda", "puede", "pueden", "puedo", "pues", "que", "quedó", "queremos", "quien", "quienes", "quiere", "quién", "qué", "realizado", "realizar", "realizó", "respecto", "sabe", "sabeis", "sabemos", "saben", "saber", "sabes", "se", "sea", "seamos", "sean", "seas", "segunda", "segundo", "según", "seis", "ser", "seremos", "será", "serán", "serás", "seré", "seréis", "sería", "seríais", "seríamos", "serían", "serías", "seáis", "señaló", "si", "sido", "siempre", "siendo", "siete", "sigue", "siguiente", "sin", "sino", "sobre", "sois", "sola", "solamente", "solas", "solo", "solos", "somos", "son", "soy", "su", "sus", "suya", "suyas", "suyo", "suyos", "sí", "sólo", "tal", "también", "tampoco", "tan", "tanto", "te", "tendremos", "tendrá", "tendrán", "tendrás", "tendré", "tendréis", "tendría", "tendríais", "tendríamos", "tendrían", "tendrías", "tened", "teneis", "tenemos", "tener", "tenga", "tengamos", "tengan", "tengas", "tengo", "tengáis", "tenida", "tenidas", "tenido", "tenidos", "teniendo", "tenéis", "tenía", "teníais", "teníamos", "tenían", "tenías", "tercera", "ti", "tiempo", "tiene", "tienen", "tienes", "toda", "todas", "todavía", "todo", "todos", "total", "trabaja", "trabajais", "trabajamos", "trabajan", "trabajar", "trabajas", "trabajo", "tras", "trata", "través", "tres", "tu", "tus", "tuve", "tuviera", "tuvierais", "tuvieran", "tuvieras", "tuvieron", "tuviese", "tuvieseis", "tuviesen", "tuvieses", "tuvimos", "tuviste", "tuvisteis", "tuviéramos", "tuviésemos", "tuvo", "tuya", "tuyas", "tuyo", "tuyos", "tú", "ultimo", "un", "una", "unas", "uno", "unos", "usa", "usais", "usamos", "usan", "usar", "usas", "uso", "usted", "va", "vais", "valor", "vamos", "van", "varias", "varios", "vaya", "veces", "ver", "verdad", "verdadera", "verdadero", "vez", "vosotras", "vosotros", "voy", "vuestra", "vuestras", "vuestro", "vuestros", "y", "ya", "yo", "él", "éramos", "ésta", "éstas", "éste", "éstos", "última", "últimas", "último", "últimos"]

In [0]:
filtro2020 = (secop
          .select("anno_firma", "valor_del_contrato","objeto_del_contrato")
          .dropna()
          .filter(secop["anno_firma"]==2020)
)

cantidad = filtro2020.count()
n = int(cantidad*0.2)

top20 = (filtro2020
             .select("objeto_del_contrato")
             .limit(n)
             .orderBy(F.desc("valor_del_contrato"))
)

tokens = (top20
          .select(F.lower(F.col("objeto_del_contrato")).alias("txt"))
          .select(F.regexp_replace("txt", r"[^\p{L}\p{N}\s]+", " ").alias("txt"))
          .select(F.split(F.col("txt"), r"\s+").alias("words"))
          .select(F.explode("words").alias("palabra"))
          .select(F.trim("palabra").alias("palabra"))
          .filter(F.length("palabra") > 0)
)

top_20_palabras = (tokens
                   .filter(~F.col("palabra").isin(stopwords))
                   .groupBy("palabra")
                   .agg(F.count("*").alias("numero"))
                   .orderBy(F.desc("numero"))
                   .limit(20)
)

display(top_20_palabras)

palabra numero servicios 62921 prestar 34222 profesionales 27278 gestión 25680 apoyo 25217 prestación 19315 actividades 14640 desarrollo 14109 salud 11190 apoyar 9559 proyecto 9419 procesos 9211 marco 9027 integral 8078 formación 7979 administrativa 7724 centro 7665 2020 7446 departamento 7053 atención 6961